fix(curriculum) Fixed test to only allow valid JSX comments for Add Comments In JSX challenge

RandellDawson opened the pull request:
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Closes #34697

I fixed the main issue discussed in #34697 which allow a user to write the comment on the same line as the opening `div` tag.  I went a little further and validate that a comment appears inside the `div` element somewhere, but still not affect the existing `h1` and `p` elements.  I also created another test and made it the 3rd to make sure the `h1` and `p` elements original content does not change, which was part of the challenge instructions' requirements.

RandellDawson committed:
fix: only allow valid jsx comments in div elem

thecodingaviator commented on curriculum/challenges/english/03-front-end-libraries/react/add-comments-in-jsx.english.md:
```diff
@@ -27,9 +27,10 @@ tests:
   - text: The <code>JSX</code> should include a comment.
-    testString: getUserInput => assert(getUserInput('index').includes('/*') && getUserInput('index').includes('*/'), 'The <code>JSX</code> should include a comment.');
-
+    testString: assert(/<div>[\s\S]*{\s*\/\*\s*\S[\s\S]*\*\/\s*}[\s\S]*<\/div>/.test(code), 'The <code>JSX</code> should include a comment.');
```
```suggestion
    testString: assert(/<div>([\s\S]*){\s*\/\*\s*\S\1\*\/\s*}\1<\/div>/.test(code), 'The <code>JSX</code> should include a comment.');
```
https://regexr.com/49ft4 is an example of these capture groups at work

thecodingaviator reviewed (dismissed):
Either I'm missing something here or something is fishy, I've modified your RegEx to use to use capture groups, an example can be seen here: https://regexr.com/49ft4 but as you can see, the text does not pass, if I put the same comment into the challenge on rocks, and it passed:
![image](https://user-images.githubusercontent.com/34807532/53715506-0147ef80-3e78-11e9-88b6-965db314bb8b.png)

RandellDawson commented:
@thecodingaviator I do not understand what the issue is which you are trying to show here.  My regex makes sure there is more than just white space in the comment.  Your JSX should fail the test, because you don't have something other than white space between the `/*` and `*/`.

Also, if I replace my regex with what you created, it would not allow a valid solution like below to pass the test.
```
const JSX = (
  <div>
    <h1>This is a block of JSX</h1>
    <p>Here's a subtitle</p>
    <div> { /* My comment */ } </div>
  </div>
);
````

Maybe I am misunderstanding the problem you are trying to point out to me.

thecodingaviator commented:
Sorry for not be precise before, what I'm saying is that a comment without text in it, is still a comment. Right? 

> My regex makes sure there is more than just white space in the comment.

I checked my suggestion, yes, I understand what you're getting at, I'm cancelling that review

RandellDawson commented:
@thecodingaviator We have a couple of choices here:

1. Rewrite the regex so that it allows an empty (only whitespace or nothing) as acceptable comments as seen below:
```jsx
{/*  */}
```
or
```jsx
{/**/}
```

2. Change the challenge instructions and test `text` to define what a valid comment would be (non-empty and not just white space).

@Manish-Giri , @ezioda004 , @moT01 What are your thoughts on the matter?

thecodingaviator commented:
How about allowing almost anything between the `/*` and `*/` because any thing between those is a comment

moT01 commented:
For clarity... the current beta allows anything (or nothing) in the comment. 

The only issue I see on beta is that it doesn't work when placing a comment after the opening div tag or the closing h1 tags, but only if it's on the same line as the tag && there's a space between the tag and comment, e.g.... 
```
const JSX = (
  <div>
    <h1>This is a block of JSX</h1>{/**/}
    <p>Here's a subtitle</p>
  </div>
);
```
works^^
```
const JSX = (
  <div>
    <h1>This is a block of JSX</h1> {/**/}
    <p>Here's a subtitle</p>
  </div>
);
```
doesn't work^^

The issue @RandellDawson mentioned about putting the comment in a new element, probably isn't that big of a deal - I doubt many campers will do that. But, the instructions don't say not to put in a new element, so for this - either allow the comment to be put in a new element via the tests, or maybe change the instructions to say "without modifying the existing h1 or p elements, or adding and elements." (or something like that) - either fix works, or doing nothing about this is probably fine too. 

As far as if nothing or whitespace should be allowed in the comment, I'm not sure - I suppose technically it's still a comment. Just an empty one.

Manish-Giri commented:
While a comment with just whitespace or nothing in it is still a comment, my opinion is not to allow such comments to pass. 

IRL, when you look at large codebases in production, you won't see many instances of empty comments. There's a high chance that if your comment is empty or has just whitespaces, it would get flagged during code review. After all, what's the use of such a comment?

As far as possible, we should be trying to teach best practices.

RandellDawson commented:
@Manish-Giri So it sounds like you would like there to be at least one character that is not a space?  If so, then I guess we should modify the instructions and the test `text` to indicate this?

ezioda004 commented:
@RandellDawson I'd prefer a solution which is as close to the real-world-passing syntax as possible and as @thecodingaviator has mentioned a comment without any text is still a valid comment.

RandellDawson commented:
OK, I will modify the regex to allow `{/**/}` at a minimum.

RandellDawson committed:
fix: allow blank comment and removed assert msgs

RandellDawson commented:
@thecodingaviator, @moT01, @Manish-Giri, @ezioda004 OK, I have now made it so that the last test accepts an empty comment like `{/**/}`.  Because of this, I changed the `text` to better reflect the expectation (I hope that is alright with everyone).

Also, I went ahead and removed the assert messages from all the `testString`s.

thecodingaviator reviewed (approved):
LGTM! 👍 Tested locally

Manish-Giri reviewed (approved):
LGTM. Thank You for your contribution.
